loader/cl_icdl.h

```diff
@@ -0,0 +1,28 @@
+#ifndef OPENCL_CL_ICDL_H
+#define OPENCL_CL_ICDL_H
+
+#include <CL/cl.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum _cl_icdl_info{
+    CL_ICDL_OCL_VERSION = 1,
+    CL_ICDL_VERSION     = 2,
+    CL_ICDL_NAME        = 3,
+    CL_ICDL_VENDOR      = 4,
+} cl_icdl_info;
+
+typedef cl_int
+(CL_API_CALL * pfn_clGetICDLoaderInfoOCLICD)(cl_icdl_info param_name,
+                                             size_t       param_value_size,
+                                             void *       param_value,
+                                             size_t *     param_value_size_ret);
```

Comment:
I'm not sure if it's too late to change the names of these enums and functions in practice, but if it's not it would be nice to align them more closely to OpenCL naming conventions:

* For the enum names, could we be a bit more descriptive?  Maybe something like `CL_ICD_LOADER_OPENCL_VERSION` instead of `CL_ICDL_OCL_VERSION`?
    * Should these enums have a suffix like other extension enums, and if so, what suffix should we use?

* For the function name, is `OCLICD` part of the function name, or is this the extension suffix?  If it is the extension suffix then it should match whatever extension suffix we choose to use for the extension enums.  If it is part of the extension name then I think it is redundant, and the name `clGetICDLoaderInfo` is sufficient on its own.

* We should use a typedef for `cl_icdl_info` (or whatever we end up calling it) vs. an enum type.  We should also `#define` the values of each enum similarly.

Reply:
As an addendum, for the type of the enum and why I didn't want to change it: the size of the enum is not guaranteed, and there may be platforms where changing to a cl_uint may break ABI. (I know this is supposed to be a theoretical problem... but still)

Reply:
Yeah, not a problem, I definitely see the value of working with shipping software so I'm personally fine with this as-is.

I do hope we can align more closely to OpenCL naming conventions if we choose to add similar ICD loader extensions in the future.  In other words, let's think of this as a one-off solution rather than establishing precedent.

I do think it would be wise to change from an enum to a fixed-size type (probably `cl_uint`) though.  As you mentioned, the size of an enum is not guaranteed, though this is determined by the compiler and not the platform - eek.

https://stackoverflow.com/questions/366017/what-is-the-size-of-an-enum-in-c